Manipulators progress

johnhaddon opened the pull request:
This PR represents all my progress to date in adding transform manipulators to Gaffer's viewer :

- Improved GLWidget overlay support
- Better support for tools in the Viewer
    - Toolbar for selecting from available tools
    - Per-tool toolbar overlays for setting tool options
- TransformTool base class for implementing transform manipulators
- Translate and Scale tools

![translatetool](https://cloud.githubusercontent.com/assets/1133871/22982683/c0c2f6da-f397-11e6-9c23-8d23a43997a8.png)

There's still work to do, the addition of a RotateTool being the most obvious thing. Since this is just a "spare time" project at the moment, I'm not sure when I'll get more time to work on it, so I'm posting it here for feedback if nothing else. I do wonder though if it might make sense to merge it anyway - some manipulators being better than none, and hopefully driving some demand to get the rest added.

andrewkaufman commented:
Do you want a full review here and note of bugs? I found a couple things so far, just playing with it for a couple minutes.

johnhaddon commented:
Yeah, feedback would be good - then I can use it as an excuse to spend a few Image Engine hours on it :)

andrewkaufman commented on python/GafferSceneUI/SelectionToolUI.py:
```diff
@@ -52,6 +52,7 @@
 		- Drag to PathFilter or Sets node to add/remove their paths
 	""",
 
+	"viewer:shortCut", "Q",
```
Is "Q" standard? I was guessing "S" for some reason. Maybe "Q" matches Maya or somthing? Can we add this to the tooltip?

andrewkaufman commented on python/GafferSceneUI/ScaleToolUI.py:
```diff
@@ -0,0 +1,52 @@
+	Tool for editing object scale.
+	""",
+
+	"viewer:shortCut", "R",
```
I suppose this is also Maya clone, but "R" for scale instead of rotate boggles my mind...

johnhaddon commented:
Made a bit more progress over the weekend :

> - If you enable TranslateTool and then deselect your node (or select an ImageNode) the Orientation plug (or maybe the whole bottom toolbar) erroneously continues to be displayed.

Fixed in 4d65f0d.

> - Possibly related, rmb on that plug's label (at IE) throws an error because `plug.ancestor( Gaffer.ApplicationRoot.staticTypeId() ) == None`

Do you have a full stacktrace for that? I don't see the error here, so suspect an IE extension may be to blame.

> - If I transform a Group (e.g. /group) and then try to transform a child (e.g. /group/cube) its incredibly confusing what is happening, regardless of orientation.

I believe I've fixed that in 0b3360f - could you check at your end?

> - No rotate? Me thinks this is a ploy to get some actual time allocated here... ok, you got me.

This PR is a ploy for sure. But the lack of a rotate is simply down to not being able to get everything I wanted done during my time off. I'll respond to your other points below - my intention is to spend some IE time on them once I've got the hash cache stuff finally done.

> - Uniform scale (e.g. ganged without actually ganging) as well.

Good point - scale's not much use without that is it? I'll get this one done.

> - Non-axis aligned (e.g. free) translation would be highly desirable as well.

I propose we defer this to a following PR, just so we can get something into people's hands sooner. I'd also like to add XY/XZ/YZ plane translate handles too.

> - Can we add a shortcut for the Crop tool? Seems to be the only one without...

Any preference?

> - If I manually gang the scale (or translate), then manipulate any of the driven axes, I get this error:

Yep, I need to be checking if the plug is settable and I'm not. This is also related to Animation - when a plug is animated, it isn't settable. What I'd like to do is to get a first release out where the error is gone but animation greys out the manipulator (yes, useless I know), and then make the two compatible. This needs a bit of thought around an AnimationAlgo API, and I don't want to delay getting manipulators out in some form any longer than necessary.

> Is "Q" standard? I was guessing "S" for some reason. Maybe "Q" matches Maya or somthing? Can we add this to the tooltip?

Yeah, I made it match Maya 'cos that's what I'm most used to. I don't know if there's another standard. I've added it to the tooltips in a20623d.

> I suppose this is also Maya clone, but "R" for scale instead of rotate boggles my mind...

There is method in the madness. Q/W/E/R are all in a row so you can swap quickly between Select/Translate/Etc without looking down.

andrewkaufman commented:
>> Possibly related, rmb on that plug's label (at IE) throws an error because `plug.ancestor( Gaffer.ApplicationRoot.staticTypeId() ) == None`

> Do you have a full stacktrace for that? I don't see the error here, so suspect an IE extension may be to blame.

Its for sure an IE extensions causing the issue (by assuming the plug with have an application ancestor). Is that not a safe assumption? I didn't actually check why that was being enforced here, just noticed that it was.

```
Traceback (most recent call last):
  File "/home/andrewk/apps/gaffer/0.31.0.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/Widget.py", line 931, in eventFilter
    return self.__contextMenu( qObject, qEvent )
  File "/home/andrewk/apps/gaffer/0.31.0.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/Widget.py", line 1111, in __contextMenu
    return widget._contextMenuSignal( widget )
RuntimeError: Exception : Traceback (most recent call last):
  File "/home/andrewk/apps/gaffer/0.31.0.0dev/cent7.x86_64/cortex/9/gaffer/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/home/andrewk/apps/gaffer/0.31.0.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/PlugValueWidget.py", line 530, in __contextMenu
    menuDefinition = self._popupMenuDefinition()
  File "/home/andrewk/apps/gaffer/0.31.0.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/PlugValueWidget.py", line 329, in _popupMenuDefinition
    self.popupMenuSignal()( menuDefinition, self )
RuntimeError: Exception : Traceback (most recent call last):
  File "/software/apps/ieRendering/8.35.0/gaffer/0.31/tools/python/2.7/cent7.x86_64/gcc/4.8.3/cortex/9/IEGafferRenderingUI/ScenePathSelectionMenu.py", line 95, in __scenePathMenuPopupMenu
    if plug.ancestor( Gaffer.ApplicationRoot.staticTypeId() ).getName() != "CaribouApp":
AttributeError: 'NoneType' object has no attribute 'getName'
```

>> If I transform a Group (e.g. /group) and then try to transform a child (e.g. /group/cube) its incredibly confusing what is happening, regardless of orientation.

> I believe I've fixed that in 0b3360f - could you check at your end?

Yep, all good now, thanks!


>> Can we add a shortcut for the Crop tool?

> Any preference?

"C"?


>> If I manually gang the scale (or translate), then manipulate any of the driven axes, I get this error:

> What I'd like to do is to get a first release out where the error is gone but animation greys out the manipulator (yes, useless I know).

That sounds like a fine first step.

>> I suppose this is also Maya clone, but "R" for scale instead of rotate boggles my mind...

> There is method in the madness. Q/W/E/R are all in a row so you can swap quickly between Select/Translate/Etc without looking down.

Yeah, I know... but I'm no professional typist and I can definitely hit T/R/S without looking down also... but matching the most prolific app in the business probably makes sense I guess... For what its worth, Houdini is S/T/R/E (and Nuke is madness).

johnhaddon commented:
> Its for sure an IE extensions causing the issue (by assuming the plug with have an application ancestor).

I suspect that error may be exposed by the current release as well then, by right clicking on the numeric fields in the ImageView toolbar.

> Is that not a safe assumption? I didn't actually check why that was being enforced here, just noticed that it was.

The Viewer makes internal nodes for the view and the tools, and since there's nowhere to parent them, it just keeps a hold of them itself. Maybe if we wanted to start saving every aspect of the layout we'd store them in the script somewhere, but for now they don't have an ancestor of any sort, so extensions need to be robust to that...

andrewkaufman commented:
None of the ImageView plugs have label widgets, so I can't reproduce it there, but I suspect you're right. It looks like that config probably doesn't need to be checking the application root anyways, its just an early out to remove some rmb options if you're not using Caribou. Lets just ignore it.

johnhaddon commented:
> If I manually gang the scale (or translate), then manipulate any of the driven axes, I get this error

Fixed in aee148c and de40392 (building on several preceding commits for disabling gadgets).

> Can we add a shortcut for the Crop tool?

Done in 3bc950e.

andrewkaufman commented on include/GafferUI/Gadget.h:
```diff
@@ -129,6 +127,17 @@ class Gadget : public Gaffer::GraphComponent
+		/// to having a disabled ancestor.
+		bool getEnabled() const;
+		/// Returns true if this gadget and all its parents up to the
+		/// specified ancestor are visible.
```
"are enabled" rather than "are visible"?

johnhaddon replied:
Fixed in 3409b19

andrewkaufman commented on src/GafferSceneUI/TranslateTool.cpp:
```diff
@@ -111,6 +112,32 @@ bool TranslateTool::affectsHandlesTransform( const Gaffer::Plug *input ) const
+		{
+			if( translation.direction[j] != 0.0f )
+			{
+				const ValuePlug *plug = selection().transformPlug->translatePlug()->getChild( j );
```
Should `plug` be child `i` instead, defined outside the `j` loop? Right now, all handles are disabled when any of them is `!settable()`.

johnhaddon replied:
I've attempted to clarify this with an additional comment in 89ef0a8. Basically, locking a single plug can cause multiple handles to be disabled when the handles aren't aligned to the axes of the transform space (because of existing rotation). Were you just querying the code or have you seen problems in practice? I think it is possible that numerical imprecision in `createTranslation()` causes the translation values in some axes to be very small non-zero numbers rather than zero, and this could cause problems. If that's the case, I think the fix will need to be in `createTranslation()` rather than here.

andrewkaufman commented:
I can't remember if we decided uniform scale would go in the PR or in a future PR. Do you recall?

johnhaddon commented:
> I can't remember if we decided uniform scale would go in this PR or in a future PR. Do you recall?

We decided it would be in this PR, but having started it I think it might make more sense to be in the RotateTool PR. For the uniform scale I've started refactoring the single Handle class into TranslateHandle, ScaleHandle etc, because as we add more and more modes (planar translate, free translate, uniform scale) to each of the tools, it's going to get more and more awkward supporting them all in a single handle type. So basically I think uniform scale will fit in more neatly with the RotateTool work than with this PR. How does that sound?

andrewkaufman commented on src/GafferSceneUI/TranslateTool.cpp:
```diff
@@ -116,6 +116,12 @@ void TranslateTool::updateHandles()
+	// space. So any given handle might affect several components
+	// of the target translation. For each handle, check to see
+	// if each of the plugs it effects are settable, and if not,
+	// disable the handle.
```
I'm willing to accept this for now, but I don't expect a user will. The simplest case (e.g. translate.x is driven by something, but I don't care, I just want to move translate.y) means I can't move anything. It kind of sounds like the enabled state of the handle should be taking the orientation plug into account?

andrewkaufman replied:
Actually, it seems my test had a rotate in it, and that was throwing off what I was expecting to see. Lets ignore me for now.

andrewkaufman replied:
Maybe having the orientation default to World would have avoided the confusion on my end. Basically, the workflow was:

- go to node graph
  - drop down a cube
- go to node editor
  - rotate in y
  - gang translate
- go to viewport
  - turn on TranslateTool
    - gah! i cant do anything
- go back to node editor
  - actually, i can do exactly what i want by scrubbing translate.x
- go back to viewport
    - flip orientation to world
    - now i can do what i want
      - and it matches what was happening when i was scrubbing in the node editor

johnhaddon replied:
> It kind of sounds like the enabled state of the handle should be taking the orientation plug into account?

I think you figured this out, but that's exactly what is happening already.

> Maybe having the orientation default to World would have avoided the confusion on my end.

Does this have the potential to add confusion when you've got a non-rotated cube with a locked translate axis inside a rotated group?

```
import Gaffer
import GafferScene
import IECore

Gaffer.Metadata.registerNodeValue( parent, "serialiser:milestoneVersion", 0, persistent=False )
Gaffer.Metadata.registerNodeValue( parent, "serialiser:majorVersion", 31, persistent=False )
Gaffer.Metadata.registerNodeValue( parent, "serialiser:minorVersion", 0, persistent=False )
Gaffer.Metadata.registerNodeValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["Cube"] = GafferScene.Cube( "Cube" )
parent.addChild( __children["Cube"] )
__children["Cube"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = IECore.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Cube"]["__uiPosition"].setValue( IECore.V2f( -11.3500004, 3.84999895 ) )
__children["Group"] = GafferScene.Group( "Group" )
parent.addChild( __children["Group"] )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["transform"]["rotate"].setValue( IECore.V3f( 0, 45, 0 ) )
__children["Group"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = IECore.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["__uiPosition"].setValue( IECore.V2f( -9.85084534, -5.31406307 ) )
Gaffer.Metadata.registerPlugValue( __children["Cube"]["transform"]["translate"]["x"], 'readOnly', True )
__children["Group"]["in"]["in0"].setInput( __children["Cube"]["out"] )


del __children
```

Maybe the default should be Parent, since it improves both situations?

andrewkaufman replied:
Ok, sure, lets try Parent. We can always change it afterwards based on some user feedback.

johnhaddon replied:
Cool - done in 60f8dbb

andrewkaufman commented:
test needs a change?

johnhaddon commented:
> test needs a change?

Gah. Fixed and repushed...

johnhaddon commented:
I've rebased to master and added support for uniform scaling. This involved a fair bit of refactoring designed to set us up for the upcoming RotateTool and the addition of planar translation to the TranslateTool. New commits start at 9f6c6d1.

andrewkaufman commented:
I'm happy to merge this as is, but if I end up waiting a day due to meeting madness, it might be worth addressing an odd behaviour of the uniform scale.

- First scale in X
- Now uniform scale
  - The proportions aren't maintained as expected. Seems to be adding rather than scaling maybe?

johnhaddon commented:
Ah yes, I've totally fluffed that haven't I? I've gone down a bit of a rabbit hole to get it fixed and I'm not quite out yet, but if it's OK I'd like to wait to have it working properly before we merge...

johnhaddon commented:
After going round the houses a bit to get a uniform scaling mechanism that felt intuitive I realised that my fixed approach using intersections-with-camera-aligned planes-with-correction-for-existing-object-scaling was exactly the same as tracking the drag in raster space. So 49f7855 and 4ce2658 now implement that much simpler approach directly, and hopefully do want you want.

It wasn't all for nothing though. The work in eb54e22 sets me up for the RotateTool and planar translation, and f3291a1 puts things on a firmer footing in terms of documentation and expected results.
